### ISSUE TITLE
Early stopping

### ISSUE THREAD

**Issue by adelmemariani** [dice-group/dice-embeddings#393]
I noticed when I reduce the number of epochs (for example from 100 to 80 or 60), usually the performance (MRR) is getting better. Therefore, I thought we should apply "early stopping", if not yet implemented.

**Comment by Demirrr** [comment]
Please do not assign me for not discussed features. We have a discussion period. Early stopping already available. Importantly there are new algorithms designed to allocate or tackle over fitting:

Stochastic Weight Averaging (SWA)
Adaptive Stochastic Weight Averaging (ASWA)
Stochastic Weight Averaging–Gaussian (SWAG)
Exponential Moving Average (EMA)

Please refer to the readme.

**Comment by adelmemariani** [comment]
I know and I also implemented some new adaptive weight averaging methods, but they are different from early stopping.

**Comment by Demirrr** [comment]
Sure. I have just double checked the code. We seem to retire the early stopping code. Early stopping can be reintegrated. Please make a pull request if you have already implementation. Please ensure that your implementation should be compatible with the dev branch.

**Comment by adelmemariani** [comment]
Thanks, we could use early stopping as a simple baseline (to compare with weight averaging methods).
I integrated the approach introduced [here](https://www.sciencedirect.com/science/article/abs/pii/S0031320325009586) to the dice embedding, but still haven't made a pull request. I'll do it next week.

**Comment by Demirrr** [comment]
Sounds good. Please also talk to @sapkotaruz11 as he was researching on the same direction

**Comment by sapkotaruz11** [comment]
@adelmemariani We currently have a feature that evaluates the model and saves checkpoints every n epochs, allowing us to capture intermediate training states and retain models at different stages—similar in purpose to early stopping, but without halting the training process. With this, you can simulate early stopping for different epochs in one training run.  However, if you have a need to stop training based on some metrics ( train/val loss or MRR ), we could as well integrate this.  Checkout [Periodic Evaluations](https://github.com/dice-group/dice-embeddings/blob/f38d9b770c45c034574ec190a71db9f57ef88dee/dicee/callbacks.py#L378)
